source/common/matcher/field_matcher.h

```diff
@@ -0,0 +1,146 @@
+
+#pragma once
+
+#include "envoy/matcher/matcher.h"
+
+namespace Envoy {
+namespace Matcher {
+
+/**
+ * The result of a field match.
+ */
+struct FieldMatchResult {
+  // Encodes whether we were able to perform the match.
+  MatchState match_state_;
+
+  // The result, if matching was completed.
+  absl::optional<bool> result_;
+
+  bool result() const { return *result_; }
+};
+
+/**
+ * Base class for matching against a single input.
+ */
+template <class DataType> class FieldMatcher {
+public:
+  virtual ~FieldMatcher() = default;
+
+  /**
+   * Attempts to match against the provided data.
+   * @returns absl::optional<bool> if matching was possible, the result of the match. Otherwise
+   * absl::nullopt if the data is not available.
+   */
+  virtual FieldMatchResult match(const DataType& data) PURE;
+};
+template <class DataType> using FieldMatcherPtr = std::unique_ptr<FieldMatcher<DataType>>;
+
+/**
+ * A FieldMatcher that attempts to match multiple FieldMatchers, evaluating to true iff all the
+ * FieldMatchers evaluate to true.
+ *
+ * If any of the underlying FieldMatchers are unable to produce a result, absl::nullopt is returned.
+ */
+template <class DataType> class AllFieldMatcher : public FieldMatcher<DataType> {
+public:
+  explicit AllFieldMatcher(std::vector<FieldMatcherPtr<DataType>>&& matchers)
+      : matchers_(std::move(matchers)) {}
+
+  FieldMatchResult match(const DataType& data) override {
+    for (const auto& matcher : matchers_) {
+      const auto result = matcher->match(data);
+
+      // If we are unable to decide on a match at this point, propagate this up to defer
+      // the match result until we have the requisite data.
+      if (result.match_state_ == MatchState::UnableToMatch) {
+        return result;
+      }
+
+      if (!result.result()) {
+        return result;
+      }
+    }
+
+    return {MatchState::MatchComplete, true};
+  }
+
+private:
+  const std::vector<FieldMatcherPtr<DataType>> matchers_;
+};
+
+/**
+ * A FieldMatcher that attempts to match multiple FieldMatchers, evaluating to true iff any of the
+ * FieldMatchers evaluate to true.
+ *
+ * If any of the underlying FieldMatchers are unable to produce a result before we see a successful
+ * match, absl::nullopt is returned.
+ */
+template <class DataType> class AnyFieldMatcher : public FieldMatcher<DataType> {
+public:
+  explicit AnyFieldMatcher(std::vector<FieldMatcherPtr<DataType>>&& matchers)
+      : matchers_(std::move(matchers)) {}
+
+  FieldMatchResult match(const DataType& data) override {
+    for (const auto& matcher : matchers_) {
+      const auto result = matcher->match(data);
+
+      // If we are unable to decide on a match at this point, propagate this up to defer
+      // the match result until we have the requisite data.
+      if (result.match_state_ == MatchState::UnableToMatch) {
```

Comment:
But why does this matter for the any case? Can't we short circuit faster if any of them are true?

Reply:
Oh I see, yeah that makes sense. I think if any of them are true we return true, but if we don't see any trues but some UnableToMatch then we return UnableToMatch